pkg/mount/mount_linux.go

```diff
@@ -2,26 +2,106 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
+// Package mount implements mounting, moving, and unmounting file systems.
 package mount
 
 import (
 	"errors"
 	"fmt"
+	"os"
 
 	"golang.org/x/sys/unix"
 )
 
+// Most commonly used mount flags.
+const (
+	MS_RDONLY   = unix.MS_RDONLY
+	MS_BIND     = unix.MS_BIND
+	MS_LAZYTIME = unix.MS_LAZYTIME
+	MS_NOEXEC   = unix.MS_NOEXEC
+	MS_NOSUID   = unix.MS_NOSUID
+	MS_NOUSER   = unix.MS_NOUSER
+	MS_RELATIME = unix.MS_RELATIME
+	MS_SYNC     = unix.MS_SYNC
+
+	MNT_FORCE  = unix.MNT_FORCE
+	MNT_DETACH = unix.MNT_DETACH
+)
+
+// Mounter is a device that can be attached at a file system path.
+type Mounter interface {
+	// Mount attaches the device at path.
+	Mount(path string, flags uintptr) (*MountPoint, error)
+}
+
+// MountPoint represents a mounted file system.
+type MountPoint struct {
+	Path   string
+	Device string
+	FSType string
+	Flags  uintptr
+	Data   string
+}
+
+// String implements fmt.Stringer.
+func (mp *MountPoint) String() string {
+	return fmt.Sprintf("MountPoint(path=%s, device=%s, fs=%s, flags=%#x, data=%s)", mp.Path, mp.Device, mp.FSType, mp.Flags, mp.Data)
+}
+
+// Unmount unmounts a file system that was previously mounted.
+func (mp *MountPoint) Unmount(flags uintptr) error {
+	if err := unix.Unmount(mp.Path, int(flags)); err != nil {
+		return &os.PathError{
+			Op:   "unmount",
+			Path: mp.Path,
+			Err:  fmt.Errorf("flags %#x: %v", flags, err),
+		}
+	}
+	return nil
+}
+
 // Mount attaches the fsType file system at path.
 //
 // dev is the device to mount (this is often the path of a block device, name
 // of a file, or a dummy string). data usually contains arguments for the
 // specific file system.
-func Mount(dev, path, fsType, data string, flags uintptr) error {
+func Mount(dev, path, fsType, data string, flags uintptr) (*MountPoint, error) {
+	// Create the mount point if it doesn't already exist.
+	if err := os.MkdirAll(path, 0666); err != nil {
+		return nil, err
+	}
+
 	if err := unix.Mount(dev, path, fsType, flags, data); err != nil {
-		return fmt.Errorf("mount(%q on %q [type %q flags %x]): %v",
-			dev, path, fsType, flags, err)
+		return nil, &os.PathError{
+			Op:   "mount",
+			Path: path,
+			Err:  fmt.Errorf("from device %q (fs type %s, flags %#x): %v", dev, fsType, flags, err),
+		}
 	}
-	return nil
+	return &MountPoint{
+		Path:   path,
+		Device: dev,
+		FSType: fsType,
+		Data:   data,
+		Flags:  flags,
+	}, nil
+}
+
+// TryMount tries to mount a device on the given mountpoint, trying in order
+// the supported block device file systems on the system.
+func TryMount(device, path string, flags uintptr) (*MountPoint, error) {
```

Comment:
This one is awesome. I may have missed it in the patch series or something but I assume cmds/core/mount will want it :-)